### PR TITLE
2-pages mode: twoVisiblePagesAsOnePageNumber, single wide header

### PR DESCRIPTION
**Pull request by poire-z**
The 3 first commits will allow more proper support for "Dual page" mode in KOReader, actually making it feel more as "Two columns".
Previous related discussions in https://github.com/koreader/koreader/issues/7129 and https://github.com/koreader/crengine/pull/408
Supercedes https://github.com/koreader/crengine/pull/408.

#### 2-pages mode: add option twoVisiblePagesAsOnePageNumber
Can be set by frontends to get one page number per view instead of 2 when in 2-pages mode, actually making it feel like 2-columns on 1 page.
Updating 4 public methods is all it needs: `getCurPage()`, `getPageCount()`, `getBookmarkPage()` and `goToPage()`. They now return or accept the "external" page number (which differs from the "internal" page number when `VisiblePageCount=2` and `m_twoVisiblePagesAsOnePageNumber=true`.
As the crengine code itself use these methods, some of them now provide internal=true to get back internal page numbers where it is needed.
This adds a bit of noise, but it should be a no-op when `m_twoVisiblePagesAsOnePageNumber` is let to be `false`. I tried to update all the things needed, even if not used by KOReader, so this should also work with CoolReader.
With KOReader, we'll probably stay with a non-user-toggable `m_twoVisiblePagesAsOnePageNumber=true`.

#### TOC, PageMap: save the visible pages number they were build for
So we can notice they are no longer valid when toggling twoVisiblePagesAsOnePageNumber (which does not change the rendering).

#### 2-pages mode: tweak header drawing
When `m_twoVisiblePagesAsOnePageNumber=true`, draw header as one single wide header like when in 1-page mode.
Also show the external current page number and pages count.

#### Header: add option to display time as 12-hours clock
No AM/PM to not have the need for them to be translated. Should allow closing https://github.com/koreader/koreader/issues/7171.

#### 2-pages mode: decrease min middle margin from 1.2 to 0.8em
Discussion and screenshots in https://github.com/koreader/koreader/discussions/7177

#### Add `LVDocView::getDocumentRenderingHash()`
So frontends can know if the rendering or the number of pages have changed (instead of relying of the document height, which is not a strictly accurate indicator of changes).

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/crengine/410)

### CODE REVIEW

File: crengine/src/lvdocview.cpp [hunk 1]
```diff
@@ -1344,7 +1344,12 @@ lString32 LVDocView::getTimeString() {
 	time_t t = (time_t) time(0);
 	tm * bt = localtime(&t);
 	char str[12];
-	sprintf(str, "%02d:%02d", bt->tm_hour, bt->tm_min);
+	if ( m_props->getBoolDef(PROP_SHOW_TIME_12HOURS, false) ) {
+		sprintf(str, "%d:%02d", bt->tm_hour > 12 ? bt->tm_hour % 12 : bt->tm_hour, bt->tm_min);
+	}
+	else {
+		sprintf(str, "%02d:%02d", bt->tm_hour, bt->tm_min);
+	}
```
Comment (NiLuJe) [inline crengine/src/lvdocview.cpp:1352]: I, err, say again: why not use `strftime`? It's ANSI C, we have no reason to reinvent the wheel here?

File: crengine/src/lvdocview.cpp [hunk 2]
```diff
@@ -2590,7 +2652,7 @@ void LVDocView::updateLayout() {
                 // left or right margin, whichever is the greatest.
 		// We still want to ensure a minimal middle margin in case
 		// the requested pageMargins are really small. Say 1.2em.
-		int min_middle_margin = 1.2 * m_font_size;
+		int min_middle_margin = 0.8 * m_font_size;
```
Comment (NiLuJe) [inline crengine/src/lvdocview.cpp:2655]: `0.8f`, we don't need `double` precision for a result that'll be truncated anyway ;).

File: crengine/src/lvdocview.cpp (hunk 1 above)
Comment (poire-z) [reply crengine/src/lvdocview.cpp:1352]: Because I have no reason to reinvent crengine and learn strftime :) (for such a minor thingie.)

File: crengine/src/lvdocview.cpp (hunk 2 above)
Comment (poire-z) [reply crengine/src/lvdocview.cpp:2655]: I want to stay with integers, as there's no (any, or not much) floats in crengine.
Should I use `m_font_size * 80 / 100` ? or does that `0.8 *` just works with integer input giving integer output? (Well, we were already using 1.2 :)

Comment (NiLuJe) [reply crengine/src/lvdocview.cpp:2655]: The integer truncation is not the issue, the implicit cast to double for the computation is ;).

When using literals, no type hint on a decimal means a `double`. You rarely need double precision, and it makes absolutely no sense here, since the result is truncated to an `integer` by the (also implicit) cast to the final storage type, `int` ;).

Comment (NiLuJe) [reply crengine/src/lvdocview.cpp:2655]: i.e., if I make every cast explicit, the current code essentially means `int margin = iceil((double) 0.8 * (double) m_font_size)`

File: crengine/src/lvdocview.cpp (hunk 1 above)
Comment (NiLuJe) [reply crengine/src/lvdocview.cpp:1352]: 12H:
`strftime(str, sizeof(str) - 1U, "%I:%M", &bt);` (append `%p` if you want the locale's AP/PM string).

24H:
`strftime(str, sizeof(str) - 1U, "%H:%M", &bt);`

File: crengine/src/lvdocview.cpp [hunk 3]
```diff
@@ -1336,7 +1344,12 @@ lString32 LVDocView::getTimeString() {
 	time_t t = (time_t) time(0);
 	tm * bt = localtime(&t);
 	char str[12];
-	sprintf(str, "%02d:%02d", bt->tm_hour, bt->tm_min);
+	if ( m_props->getBoolDef(PROP_SHOW_TIME_12HOURS, false) ) {
+		strftime(str, sizeof(str), "%l:%M", bt);
+	}
+	else {
+		strftime(str, sizeof(str), "%H:%M", bt);
```
Comment (NiLuJe) [inline crengine/src/lvdocview.cpp:1351]: If you want the same behavior as `%l` for a 24h clock, that's `%k` ;).

Both are somewhat non-portable, though, c.f., https://github.com/koreader/koreader/pull/6973/commits/46ebbc5825b4e46c1d31f4c0befbe5d3fd63e57a

(There must have been a reason why I went with the dash modifier instead of those, too, probably BSD or Android related).

Comment (NiLuJe) [reply crengine/src/lvdocview.cpp:1351]: (The previous code enforced %H behavior, FWIW).

File: crengine/src/lvdocview.cpp [hunk 4]
```diff
@@ -2589,8 +2651,8 @@ void LVDocView::updateLayout() {
                 // We will ensure a max middle margin the size of a single
                 // left or right margin, whichever is the greatest.
 		// We still want to ensure a minimal middle margin in case
-		// the requested pageMargins are really small. Say 1.2em.
-		int min_middle_margin = 1.2 * m_font_size;
+		// the requested pageMargins are really small. Say 0.8em.
+		int min_middle_margin = m_font_size * 80 / 100;
```
Comment (NiLuJe) [inline crengine/src/lvdocview.cpp:2655]: That's not any better ;p. The compiler will (hopefully) fold that into 0.8 at compile-time, but that's still a double ;).

File: crengine/src/lvdocview.cpp (hunk 3 above)
Comment (poire-z) [reply crengine/src/lvdocview.cpp:1351]: There was no previous code for 12-hours clock. And my -12 made it behave like %l.
I think 9:15 makes it look like it's a 12-hours clock. With 09:15, I would expect a 24-hours clock.
Not touching the 24h format, as it's what in CoolReader and its .ru users might prefer that (dunno, don't care about this stupid top status bar...)

File: crengine/src/lvdocview.cpp (hunk 4 above)
Comment (poire-z) [reply crengine/src/lvdocview.cpp:2655]: It's better looking. Then, the compiler can do what it wants... (and we do it that way so many places - and this one is actually called so rarely that I don't mind.)

File: crengine/src/lvdocview.cpp (hunk 3 above)
Comment (NiLuJe) [reply crengine/src/lvdocview.cpp:1351]: It'll break the build on Win32, though, in case it matters for CRe (or, well, us, unfortunately :s).

File: crengine/src/lvdocview.cpp (hunk 4 above)
Comment (NiLuJe) [reply crengine/src/lvdocview.cpp:2655]: Fair enough ;).

Comment (NiLuJe) [reply crengine/src/lvdocview.cpp:2655]: Actually, that got me curious ;).

Since everything involved is an `int`, it *does* coerce the compiler into purely euclidean division.

(Of course, a MUL  and a DIV is worse than a MUL to a double or a float, but, oh, well ;p).

File: crengine/src/lvdocview.cpp (hunk 3 above)
Comment (poire-z) [reply crengine/src/lvdocview.cpp:1351]: > It'll break the build on Win32

?! And the previous tm_hour stuff wouldn't ? Of course it matters, they build for windows.
Should I no longer listen to your nerdy suggestions ? :/

Comment (NiLuJe) [reply crengine/src/lvdocview.cpp:1351]: I recommended `%I` for a reason ;p.

Comment (poire-z) [reply crengine/src/lvdocview.cpp:1351]: So, if I want 9:15 on Windows, I just have to get back to the original sprintf(), right?
Simpler is better is calling it quit.

Comment (NiLuJe) [reply crengine/src/lvdocview.cpp:1351]: Yup.

c.f., [musl](https://git.musl-libc.org/cgit/musl/tree/src/time/strftime.c#n98) for an implementation that doesn't use DIVs ;).

**Review by NiLuJe — APPROVED** [review]

**Comment by virxkane** [comment]
@poire-z Just a note.
For now 2-page mode is transformed into 2-column mode. But this was done rather ugly, even more crutches were added into engine, it would be more correct to modify LVPageSplitter. Of course, this is much more difficult.
I don't even know what to do :)
In addition, there are disadvantages of such a 2-column mode:
Here is a screenshot from KOReader 2021.05 "Yellow Daffodil", book is the same "Proffer_Klyuchi-k-Lolite-.EeRvzA.159923.fb2":
Screenshot_1622448305_sz
The title of the book is not positioned correctly, it should be centered on the page, not in the column.
Book title definition in source file:
```xml
<body>
  <title>
   <p><strong>Карл Проффер</strong></p>
   <p><strong>Ключи к "ЛОЛИТЕ"</strong></p>
  </title>
  <epigraph>
   <p>for Ellendea</p>
  </epigraph>
...
</body>
```
May be other top level headers also must be centered on page, but low level header are ok.

**Comment by poire-z** [comment]
> For now 2-page mode is transformed into 2-column mode. But this was done rather ugly, even more crutches were added into engine

Well, you also added your own part of crutches :) larger chapter marks, option to display status bar at bottom, that I didn't pick :)

> I don't even know what to do :)

I think we have diverted too much on this header/status bar, we each added stuff that we needed, and don't care much about what the other side added (and, as you I guess, this being not essential, and this part of lvdocview.cpp being already ugly from the start, I don't really want to spend much time understanding what the other side has done :)

> it would be more correct to modify LVPageSplitter. Of course, this is much more difficult.

Well, nothing to do with LVPageSplitter : this does not change anything to page rendering and splitting - and I don't want this to have to touch anything related.
This just changes how page numbers and total nb of pages are reported to frontends. So, 2-columns looks exactly like 2-pages, but it reports being a single page number.
I tried to make it optionnal so it should cause no issue on your side if you don't enable this feature.

> The title of the book is not positioned correctly, it should be centered on the page, not in the column.

I have zero plan to change anything about this.
But I guess you have also this when in 1-page or 1-column mode: on the first page, the title is not centered in the first page.
It's a FB2 styling issue I guess - there's no reason to center anything and not respect publishers' top margin or its absence in other formats.
For FB2, as we are half the publisher as we set the book style, may be just adding this to fb2.css could make you happy?:
`body > title { margin-top: 40vh} ` which means 40% of viewport height

image

**Comment by poire-z** [comment]
If you meant the title page should take a full screen view, so having pages made of 2 columns mixed with page with a single column, well, that's a no-go for me :) Too much changes needed for a title page people will read for 5 seconds vs 10 hours the rest of the book :)
Previous discussions with users about the choices made at https://github.com/koreader/koreader/issues/7129#issuecomment-763602579.

**Comment by virxkane** [comment]
> Well, you also added your own part of crutches :) larger chapter marks, option to display status bar at bottom, that I didn't pick :)

Here I may have expressed myself incorrectly. By the crutch, I mean the so-called _dirty hack_.
Yes, and these are not crutches, but features. The option to display the status bar at the bottom of the screen is useful on some eink devices, where there is uneven brightness in the backlight just at the bottom of the screen. If you move the status bar down, this drawback becomes less noticeable.
Yes, it seems that in KOReader the status bar is drawn in the frontend, but this is an optional feature and would not affect you in any way. Large chapter marks are actually not large - at dpi = 96 there are practically no changes, at increasing dpi - it is scaled accordingly. It seemed to me so more correct.
But you know, I never insist on my option, it's up to you.

Hmm, you forgot about subpixel font rendering :)

> I think we have diverted too much on this header/status bar, we each added stuff that we needed, and don't care much about what the other side added

Yes, so be it.

> (and, as you I guess, this being not essential, and this part of lvdocview.cpp being already ugly from the start, I don't really want to spend much time understanding what the other side has done :)

Yes, it happens, when creating a program, some functions were thought about, skeleton was laid for them, then they were implemented. But later, in order to add new functions, sometimes it turns out that the existing architecture does not allow this. Therefore, we must change the architecture of the program (and, as a result, rewrite the already implemented functions) or make crutches around the existing code, making it more and more complicated, confusing, illogical and, as a result, greatly increasing the entry threshold. Personally, I don't like this approach, and I think you too, but sometimes we have to make compromises and I also sometimes have to make crutches.

> Well, nothing to do with LVPageSplitter : this does not change anything to page rendering and splitting - and I don't want this to have to touch anything related.
This just changes how page numbers and total nb of pages are reported to frontends. So, 2-columns looks exactly like 2-pages, but it reports being a single page number.
I tried to make it optionnal so it should cause no issue on your side if you don't enable this feature.

It seems to me that LVPageSplitter should support 2-column mode, but I haven't learn it and I don't know how difficult it is. Now in your implementation this is not a real 2-column mode - it is a hacked 2-page mode. The functions you added `getExternalPageNumber()`, `getInternalPageNumber()`, the added `internal` argument in the functions `getBookmarkPage()`, `getCurPage()`, `goToPage()`, `getPageCount()` - this is, in my opinion, a crutch or if you want a hack.
Ideally, we must add support for 2-column mode to `LVPageSplitter`, and then there will be no need for shenanigans around `LVDocView::m_pages` and related stuff, one page - one `m_pages` element, which contains two columns. But I certainly do not insist.

> I have zero plan to change anything about this.

No problems, everyone has their own priorities.

> But I guess you have also this when in 1-page or 1-column mode: on the first page, the title is not centered in the first page.
It's a FB2 styling issue I guess - there's no reason to center anything and not respect publishers' top margin or its absence in other formats.
For FB2, as we are half the publisher as we set the book style, may be just adding this to fb2.css could make you happy?:
body > title { margin-top: 40vh} which means 40% of viewport height

I meant a little different: the title of the book should be centered horizontally to the full width of the screen, not the column.

In general, I appreciate your work, but a little healthy criticism it's good, right? :)
About a year ago, when I continued to transfer your work to CoolReader after pkb, you asked to report any errors found. This is certainly not error, but, in my opinion, drawback, and I reported it. :)
Anyway, thanks for your work.

**Comment by virxkane** [comment]
> Too much changes needed for a title page people will read for 5 seconds vs 10 hours the rest of the book :)

But this does not prevent us from doing other things, which people see for 5 seconds, but we do it anyway, maybe because otherwise it is not pleasant to look :)

**Comment by poire-z** [comment]
> Yes, and these are not crutches, but features. 
> But you know, I never insist on my option, it's up to you.

I don't doubt these are features. But in the case of lvdocview.cpp, and these status/header parts, I have zero interest in them, and the code is ugly (mostly because of the mixed indentation that makes it unreadable/unwritable to me - and again, having no interest, I don't want to fix that indentation and mess git history, because having no interest, git history is my only help when fixing little stuff.) (Did I mention I have no interest in the header code ? :)
I don't use the top status bar, and I don't use 2-pages/2-columns mode.
I was just fixing the inconsistency of page numbers in frontend for the interested users, trying to make it the less intrusive possible - and getExternalPageNumber()/getInternalPageNumber() is exactly what I wanted to solve :) (Notice that this PR does not mention "2 columns", only frontend does, just because 2 book pages on a single screen with a single page number makes the name "2 columns" more appropriate than anything else.)
I really have no personnal interest in a more "correct" 2-columns mode - and what we get with the existing 2-pages mode + these tweaks to page numbers made our users happy enough - so, that's a done work for me :)

You can consider it a dirty hack, but that's because you read more that what's on the tin of this PR :)

> It seems to me that LVPageSplitter should support 2-column mode

I guess it could (but you may not see yet the work needed elsewhere to support that, like highlighting that we do some part in frontend). But no personal interest spending time on it, now that the above is good enough.

> In general, I appreciate your work, but a little healthy criticism it's good, right? :)

No problem :) But I wouldn't even call this PR "my work" :) I was just solving a KOReader issue while making the changes non-intrusive _for you_ when you pick it: you don't need any frontend change, you get your current behaviour while we can get our expected behaviour with a new property. That's the only "work" I expect you to appreciate from this PR, nothing else :).

> Hmm, you forgot about subpixel font rendering & color fonts :)

Yes, I explicitely forgot it :) If the involved code weren't that different and it was easy merging, I would just have picked it. But with the differences, it's probably a few days work, and LCD and color emoji are not our first target audience, eInk is :) (Also, it's quite localized, so I feel it shouldn't hurt later if I don't have it, when I pick other interesting stuff from you.) (Also, it was a lot of buffer drawing tweaks, and that's not some part of the code I'm familiar with - but @NiLuJe is :) that's https://github.com/buggins/coolreader/pull/289 if there's some interest :)

**Comment by virxkane** [comment]
> I really have no personnal interest in a more "correct" 2-columns mode - and what we get with the existing 2-pages mode + these tweaks to page numbers made our users happy enough - so, that's a done work for me :)

Ok.

> I guess it could (but you may not see yet the work needed elsewhere to support that, like highlighting that we do some part in frontend).

It's good, may be someone implement this, but... who needs it? :)

> But no personal interest spending time on it, now that the above is good enough.

I got what you mean, no problem.

> If the involved code weren't that different and it was easy merging, I would just have picked it.

No, it's will be not easy :)

**Comment by NiLuJe** [comment]
FWIW, I was going to write a lengthy post explaining why we don't really want to deal with SPR, and then @inklesspen posted something on Discord that nicely sums it up (and more):

So: https://gankra.github.io/blah/text-hates-you/#subpixel-aa-isnt-composable

TL;DR: It's complicated, and unnecessary on high-dpi devices, which pretty much includes everything we target (e.g., e-Ink & not-completely-terrible Android devices); and as far as Android is concerned, it's horribly hard and panel-specific to do on OLED because of their *many* funky pixel layouts (usually thrown under the umbrella 'Pentile', but there are many subtleties, and things are even worse on OLED with white subpixels).

(I'm not aware of anything trying to handle this properly on OLED via FreeType, which only provides basic RGB/BRG H/V filters).

**Comment by virxkane** [comment]
> @inklesspen posted something on Discord that nicely sums it up (and more, highly recommended reading):
> 
> So: https://gankra.github.io/blah/text-hates-you/#subpixel-aa-isnt-composable

In general, it is right, except that we do not have any animation.

> It's complicated, and mostly unnecessary on high-dpi devices

Yes, of course, but, for example, CoolReader is often used on a desktop.

> I'm not aware of anything trying to handle this properly on OLED via FreeType, which only provides basic RGB/BGR H/V filters

FreeType with Harmony subpixel LCD rendering also support PenTile. You can set absolutely any pixel configuration, but it's difficult. In addition, I saw that, for example, on the Galaxy A3 (2016) /AMOLED/, subpixel anti-aliasing is already implemented either at the system level or at the hardware level, so this is most likely not necessary at the software level.
But on IPS displays with RGB subpixel structure it looks great.

**Comment by NiLuJe** [comment]
Huh, I'd missed `FT_Library_SetLcdGeometry` when harmony landed ;).

In any case, yeah, it's great on low dpi screens, i'm not disagreeing there: we just don't really target that ^^.

File: crengine/src/lvdocview.cpp (hunk 4 above)
Comment (zwim) [reply crengine/src/lvdocview.cpp:2655]: Sorry for calling this old thing again. If the concerns about division still exists, instead of `80/100` some can do `819/1024` which yields 0.799805. Here the compiler generates a bitshift by 10 only, no division.
Maybe @poire-z  wants to include this the next time?

Comment (poire-z) [reply crengine/src/lvdocview.cpp:2655]: No real need: the compiler will do an int division, and that's just fine ; it's one (more) division done only when re-rendering, so we don't need to optimize anything.
And `819/1024` would need some comment explaining why this odd combination :)

(This discussion was about trying to not do float arithmetic, and we end up being convinced there was not float around here.)

Comment (poire-z) [reply crengine/src/lvdocview.cpp:2655]: From https://github.com/koreader/koreader/issues/8440, I assume you're happy with 80/100 once you have removed the publisher's own added/excessive margins, right ?
(I hope so, I don't want to go into https://github.com/koreader/koreader/discussions/7177 all over again :)

**Comment by zwim** [comment]
You are right. Style tweak did the magic.

Am 12. November 2021 23:19:41 MEZ schrieb poire-z ***@***.***>:
***@***.*** commented on this pull request.
>
>
>
>> @@ -2589,8 +2651,8 @@ void LVDocView::updateLayout() {
>                 // We will ensure a max middle margin the size of a single
>                 // left or right margin, whichever is the greatest.
> 		// We still want to ensure a minimal middle margin in case
>-		// the requested pageMargins are really small. Say 1.2em.
>-		int min_middle_margin = 1.2 * m_font_size;
>+		// the requested pageMargins are really small. Say 0.8em.
>+		int min_middle_margin = m_font_size * 80 / 100;
>
>From https://github.com/koreader/koreader/issues/8440, I assume you're happy with 80/100 once you have removed the publisher's own added/excessive margins, right ?
>(I hope so, I don't want to go into https://github.com/koreader/koreader/discussions/7177 all over again :)
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/koreader/crengine/pull/410#discussion_r748617169
-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.
